src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -0,0 +1,378 @@
+
+  def genotypeType: Type = metadata.genotypeSignature
+
+  def typ = TStruct(
+    "v" -> TVariant,
+    "va" -> vaType,
+    "s" -> TString,
+    "sa" -> saType,
+    "g" -> genotypeType)
+
+  def sampleEC: EvalContext = {
+    val aggregationST = Map(
+      "global" -> (0, globalType),
+      "s" -> (1, TString),
+      "sa" -> (2, saType),
+      "g" -> (3, genotypeType),
+      "v" -> (4, TVariant),
+      "va" -> (5, vaType))
+    EvalContext(Map(
+      "global" -> (0, globalType),
+      "s" -> (1, TString),
+      "sa" -> (2, saType),
+      "gs" -> (3, TAggregable(genotypeType, aggregationST))))
+  }
+
+  def variantEC: EvalContext = {
+    val aggregationST = Map(
+      "global" -> (0, globalType),
+      "v" -> (1, TVariant),
+      "va" -> (2, vaType),
+      "g" -> (3, genotypeType),
+      "s" -> (4, TString),
+      "sa" -> (5, saType))
+    EvalContext(Map(
+      "global" -> (0, globalType),
+      "v" -> (1, TVariant),
+      "va" -> (2, vaType),
+      "gs" -> (3, TAggregable(genotypeType, aggregationST))))
+  }
+}
+
+object NewAST {
+  def genericRewriteTopDown(ast: NewAST, rule: PartialFunction[NewAST, NewAST]): NewAST = {
+    def rewrite(ast: NewAST): NewAST = {
+      rule.lift(ast) match {
+        case Some(newAST) if newAST ne ast =>
+          rewrite(newAST)
+        case None =>
+          val newChildren = ast.children.map(rewrite)
+          if ((ast.children, newChildren).zipped.forall(_ eq _))
+            ast
+          else
+            ast.copy(newChildren)
+      }
+    }
+
+    rewrite(ast)
+  }
+
+  def rewriteTopDown[T](ast: MatrixAST[T], rule: PartialFunction[NewAST, NewAST]): MatrixAST[T] =
+    genericRewriteTopDown(ast, rule).asInstanceOf[MatrixAST[T]]
+
+  def rewriteTopDown[T](ast: KeyTableAST, rule: PartialFunction[NewAST, NewAST]): KeyTableAST =
+    genericRewriteTopDown(ast, rule).asInstanceOf[KeyTableAST]
+
+  def genericRewriteBottomUp(ast: NewAST, rule: PartialFunction[NewAST, NewAST]): NewAST = {
+    def rewrite(ast: NewAST): NewAST = {
+      val newChildren = ast.children.map(rewrite)
+
+      // only recons if necessary
+      val rewritten =
+        if ((ast.children, newChildren).zipped.forall(_ eq _))
+          ast
+        else
+          ast.copy(newChildren)
+
+      rule.lift(rewritten) match {
+        case Some(newAST) if newAST ne rewritten =>
+          rewrite(newAST)
+        case None =>
+          rewritten
+      }
+    }
+
+    rewrite(ast)
+  }
+  
+  def rewriteBottomUp[T](ast: MatrixAST[T], rule: PartialFunction[NewAST, NewAST]): MatrixAST[T] =
+    genericRewriteBottomUp(ast, rule).asInstanceOf[MatrixAST[T]]
+
+  def rewriteBottomUp[T](ast: KeyTableAST, rule: PartialFunction[NewAST, NewAST]): KeyTableAST =
+    genericRewriteBottomUp(ast, rule).asInstanceOf[KeyTableAST]
+}
+
+abstract class NewAST {
+  def children: IndexedSeq[NewAST]
+
+  def copy(newChildren: IndexedSeq[NewAST]): NewAST
+
+  def mapChildren(f: (NewAST) => NewAST): NewAST = {
```

Comment:
I would just call this `map`. `NewAST` would be a true functor if it's recursive references to `NewAST` were references to some parameterized type:

```scala
trait NewAST[T] {
  def map(f: T => U): NewAST[U]
}
case class FilterSamples[T](child: T, pred: AST) extends NewAST[T] { ... }
```

You wouldn't want to actually implement this because you'd need some allocation inducing fix point operator:

```scala
case class Mu[F[_]](unMu: F[Mu[F]]) { }
```

So that you could write the least fixed-point type:

```
type RecrusiveNewAST = Mu[NewAST]
```

But at it's heart `mapChildren` is just a restricted functorial `map`.